libsolidity/interface/CompilerStack.cpp

```diff
@@ -841,3 +842,81 @@ string CompilerStack::computeSourceMapping(eth::AssemblyItems const& _items) con
 	}
 	return ret;
 }
+
+namespace {
+
+Json::Value gasToJson(GasEstimator::GasConsumption const& _gas)
+{
+    if (_gas.isInfinite || _gas.value > std::numeric_limits<Json::LargestUInt>::max())
+	return Json::Value(Json::nullValue);
+    else
+	return Json::Value(Json::LargestUInt(_gas.value));
+}
+
+}
+
+Json::Value CompilerStack::gasEstimates(string const& _contractName) const
+{
+	if (!assemblyItems(_contractName) && !runtimeAssemblyItems(_contractName))
+		return Json::Value();
+
+	using Gas = GasEstimator::GasConsumption;
+	Json::Value output(Json::objectValue);
+
+	if (eth::AssemblyItems const* items = assemblyItems(_contractName))
+	{
+		Gas executionGas = GasEstimator::functionalEstimation(*items);
+		u256 bytecodeSize(object(_contractName).bytecode.size());
+		Gas dataGas = bytecodeSize * eth::GasCosts::createDataGas;
+
+		Json::Value creation(Json::objectValue);
+		creation["dataCost"] = gasToJson(dataGas);
+		creation["executionCost"] = gasToJson(executionGas);
+		/// TODO: implement + overload to avoid the need of +=
+		executionGas += dataGas;
+		creation["totalCost"] = gasToJson(executionGas);
+		output["creation"] = creation;
+	}
+
+	if (eth::AssemblyItems const* items = runtimeAssemblyItems(_contractName))
+	{
+		/// External functions
+		ContractDefinition const& contract = contractDefinition(_contractName);
+		Json::Value externalFunctions(Json::objectValue);
+		for (auto it: contract.interfaceFunctions())
+		{
+			string sig = it.second->externalSignature();
+			externalFunctions[sig] = gasToJson(GasEstimator::functionalEstimation(*items, sig));
+		}
+
+		if (contract.fallbackFunction())
+			externalFunctions[""] = gasToJson(GasEstimator::functionalEstimation(*items, ""));
```

Comment:
Please keep `INVALID`. The fallback function is called for an invalid signature. There is an optimization that jumps directly to the fallback if the input is empty, so I think it would take this optimization here and thus not calculate the upper bound correctly.